fix log for notifier(instance) without __name__

obarisk opened the pull request:
fix [#41563](https://github.com/apache/airflow/issues/41563)

<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

royzhong7070 committed:
fix log for notifier(instance) without __name__

royzhong7070 committed:
fix: test

josix commented on airflow/models/taskinstance.py:
```diff
@@ -1550,11 +1550,21 @@ def _run_finished_callback(
     if callbacks:
         callbacks = callbacks if isinstance(callbacks, list) else [callbacks]
```
Would you consider having a function to handle the logic behind the process of getting names for different instances? With it, we could simply call the function to get the name instead of checking the attributes every time when accessing the `__name__`.

```suggestion
        callbacks = callbacks if isinstance(callbacks, list) else [callbacks]
        def get_callback_name(callback):
            if hasattr(callback, "__name__"):
                return callback.__name__
            elif hasattr(callback, "__class__"):
                return callback.__class__.__name__
            else:
                return str(callback.__code__)
```

uranusjr commented on airflow/models/taskinstance.py:
```diff
@@ -1550,11 +1550,21 @@ def _run_finished_callback(
+            elif hasattr(callback, "__class__"):
+                log.info("Executing %s callback", callback.__class__.__name__)
+            else:
+                log.info("Executing %s callback", callback.__code__)
```
Honestly, why do we not just do `log.info("Executing %s callback", callback)`?

josix replied:
+1, I'm unsure what is the context here, seems @romsharon98 has tried it before[[link](https://github.com/apache/airflow/pull/38892#discussion_r1566141583)], do we expect the actual name of a callable instead of `repr` one

obarisk replied:
yes callback itself provides the good enough information.
Perhaps we are much like to the "position" instead of actual name.

Take the following example

if we have

```python
on_success_callback = [
  slack1,
  slack2,
  slack3
]
```

with only "name" we can't distinguish which "slack" failed.

uranusjr replied:
Yeah position in list + the function object itself sounds like the best approach to me.

royzhong7070 committed:
fix: tests

obarisk committed:
fix: tests

uranusjr committed:
Pass in entire callback object as fallback

Logging can convert an arbitrary object into a string by itself, so we
don't need to call str() eagerly. This can be a bit faster, and more
importantly, preserve as much context as possible for the logging stack
to extract if needed.

josix reviewed (approved):
Learn about `contextlib.suppress`, thanks @uranusjr

obarisk committed:
fix: test

uranusjr commented on tests/models/test_taskinstance.py:
```diff
@@ -3421,14 +3422,27 @@ def on_execute_callable(context):
+        def on_finish_callable(context):
+            pass
+
+        for callback_input in [[on_finish_callable, on_finish_callable], on_finish_callable]:
```
Let’s use `pytest.mark.parametrize` instead of a for loop

obarisk replied:
let me check the api

obarisk replied:
I'll check parametrize later.
But add enumerate won't fix the test.

Instead, need to move stdout recorder out of the loop.

obarisk committed:
fix: test

obarisk committed:
fix: test

obarisk committed:
fix: test
